### ISSUE TITLE
istio-proxy doesn't close downstream connection when TCP keepalives fail

### ISSUE THREAD

**Issue by dimara** [istio/istio#32116]
**Summary**

With Istio 1.9 installed, and TCP keepalives enabled globally, when the connection with the upstream server dies ungracefully, we notice that after all TCP keepalive probes fail, istio-proxy closes the upstream connection **but** the workload's downstream connection remains ESTABLISHED. As a result the workload hangs.

**Bug description**

We have a workload that watches the K8s API for new Pod events. The workload runs with an Istio sidecar. We have seen the API server to close the connection when it is idle for a while. To keep the connection "busy", and avoid implementing TCP keepalives in all of our workloads, we decided to enable TCP keepalives globally on Istio. On a healthy system, with an idle connection, we are seeing one probe every *time* seconds and the corresponding ACK. Still, detecting dead connections does not work as expected. Specifically:
1. The upstream server dies
2. After *time* goes by, TCP considers the connection idle and starts sending TCP keepalive probes every *interval*.
3. Since the upstream connection is dead, no TCP keepalive ACK get received.
4. After all probes fail, `istio-proxy` closes the upstream connection but leaves the downstream connection intact.
6. Our workload hangs with an ESTABLISHED connection with istio-proxy.

[ ] Docs
[ ] Installation
[X] Networking
[ ] Performance and Scalability
[ ] Extensions and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure
[ ] Upgrade

**Expected behavior**

After all TCP keepalive probes fail, istio-proxy should close the downstream connection as well and the script should raise an Exception.

**Steps to reproduce the bug**

To reproduce a problematic environment we will:
* use a watch on K8s API server as a workload with an idle outbound connection
* drop all Pod's incoming traffic, to emulate a broken connection (so that the TCP keepalives ACK sent by the API server will not get received)

Here are the steps to reproduce the issue:

1. Configure Istio TCP keepalives with time=20s, probes=3, interval=5s.
2. Enable istio sidecar injection for the default namespace:
   ```
   kubectl label namespace default istio-injection=enabled --overwrite
   ```
3. Deploy a **privileged** pod and see that it gets a sidecar.
4. Start a Python script with a simple watch:
   ```
   import kubernetes
   kubernetes.config.load_incluster_config()
   w = kubernetes.watch.Watch()
   v1 = kubernetes.client.CoreV1Api()
   stream = w.stream(v1.list_namespaced_pod, "default")
   for e in stream:
       pod_name = e["object"].metadata.name
       print("Got event for pod %s" % pod_name)
   ```
5. Monitor traffic with the API server::
   ```
   tcpdump -i any -n -vvv -x -ttt port 443
   ```
6. Monitor connections with the API server:
   ```
   watch "netstat -nap | grep 443"
   ```
7. Drop incoming traffic:
   ```
   iptables -I INPUT -j DROP
   ```
8. Wait all probes to fail, i.e., time + probes * interval = 20s + 3*5s = 35sec.
9. The script hangs.

After (4) we see **two** connections to the API server:
```
tcp        0      0 10.244.1.15:33438       20.61.26.153:443        ESTABLISHED -
tcp        0      0 10.244.1.15:41968       10.0.0.1:443            ESTABLISHED 3268/python
```
Istio "highjacks" our outbound connection and Istio becomes a proxy. So we end up with a workload->apiserver connection and an istio-proxy->apiserver one.

After (8) we noticed the TCP keepalive probes as expected.
After (9) Istio closes **its** connection with the API server (upstream) but the workload's connection remains ESTABLISHED (downstream).

**Version** 
```
# istioctl version --remote
client version: 1.9.0
control plane version: 1.9.0
data plane version: 1.9.0 (3 proxies)
```
```
# kubectl version --short
Client Version: v1.17.17
Server Version: v1.17.16
```

**How was Istio installed?**
Istio 1.9 demo profile + istioctl manifests generate.

**Environment where the bug was observed (cloud vendor, OS, etc)**
K8s 1.17

**Comment by howardjohn** [comment]
Is the upstream service TCP or HTTP (per https://istio.io/latest/docs/ops/configuration/traffic-management/protocol-selection/)?

**Comment by howardjohn** [comment]
I guess if its kubernetes.default.svc its probably `https`?

**Comment by dimara** [comment]
Hey @howardjohn. Thanks for taking a look at this.

> I guess if its kubernetes.default.svc its probably https?

Correct. The upstream service is `kubernetes` which means `https`.

**Comment by dimara** [comment]
@howardjohn @lambdai @ramaraochavali, after digging into this a bit more, I discovered that this bug report is a false alarm. 

The way I tried to reproduce the problematic environment was to drop all incoming traffic using `iptables -I INPUT -j DROP` (see step 7). Unfortunately this drops not only the apiserver->envoy traffic but also the envoy->workload one. As a result the FIN packet that envoy was sending to my workload was never received and the connection remained ESTABLISHED. I tried the above scenario by dropping the incoming traffic **only** on `eth0`, i.e., `iptables -I INPUT -i eth0 -j DROP` and everything works as expected. All connections get closed and my workload exits with:
```
urllib3.exceptions.ProtocolError: ('Connection broken: IncompleteRead(0 bytes read)', IncompleteRead(0 bytes read))
```
Sorry again for the false alarm. Closing issue.
